### PR TITLE
Revamped bump up jsoo and alcotest and dune

### PR DESCRIPTION
**Pull request by dannywillems**
TL;DR: in draft because of [this comment](https://github.com/o1-labs/snarkyjs/pull/1056/commits/edb8056b185c475872db6790e4d148499c70c5d4) in snarkyjs. On hold atm.

Bumping up js_of_ocaml to 5.4.0, dune to 3.9.2 and alcotest to 1.7.0. Cherry-pick a commit from #13790 for consistency.

Since alcotest.1.5.0, alcotest-js is available.
A better integration of node has been made in dune >= 3.8.0 (I don't remember exactly the version).
And upgrading js_of_ocaml to revive https://github.com/MinaProtocol/mina/pull/12978

It will be useful for crypto team to write tests running in JS and in native, see https://github.com/MinaProtocol/mina/issues/13537.
If a library has alcotest tests, we can activate "for free" the jsoo tests by adding `(modes native js)` in the `tests` definition. For instance, it is activated for base58_check in [this commit](https://github.com/MinaProtocol/mina/pull/13777/commits/5a47db2ae1cca21db612ba87a9f1ba0bb1b82fc3). It will run automatically a new executable. It is executed in a node environment (dune uses the `node' executable available in the path). If needed, for instance to load wasm first, the node binary can be overwritten with:
```
(env
 (_
  ; wrapping node
  (binaries
   (%{workspace_root}/test/node_wrapper.exe as node))))
```
where `node_wrapper` is an OCaml program.

Related snarkyjs and snarkyjs-bindings PR: https://github.com/o1-labs/snarkyjs/pull/1056 and https://github.com/o1-labs/snarkyjs-bindings/pull/102.

Current state:
- [x] waiting for new CI on develop - git merge when it is done.
- [x] Update new docker images.
- [x] digestif must be updated to 1.1.4, see https://github.com/mirage/digestif/releases/tag/v1.1.4. Updating implies a change in `pickles/ro.ml`
- [x] Change ledger from a class to an object. Failing tests on node because maybe top level evaluation.
```
 finished build
Running basic integration tests
/workspace_root/src/lib/snarkyjs/src/bindings/ocaml/overrides.js:34
    if (err instanceof Error) throw err;
                              ^


TypeError: Cannot read properties of undefined (reading 'apply')
    at Function.Class.<computed> (/home/runner/work/snarkyjs/snarkyjs/src/bindings/js/proxy.js:20:52)
    at Module.LocalBlockchain (/home/runner/work/snarkyjs/snarkyjs/src/lib/mina.ts:385:25)
    at file:///home/runner/work/snarkyjs/snarkyjs/src/examples/zkapps/hello_world/run.tmp.js:96:18
```
- [x] base58 not iterable
- [ ] https://github.com/o1-labs/snarkyjs/actions/runs/5712998808/job/15477548408?pr=1056. Unknown yet. **Blocked.**
- [x] Base58 undefined in base58.unit-test.ts in snarkyjs. Fixed by changing from string to bytes. Let's be careful about that!
- [x] `src/lib/base58_check`: js alcotest tests do not work. Fixed by using `digestif.ocaml`
- [ ] Update [this comment](https://github.com/o1-labs/snarkyjs-bindings/pull/79/files#diff-73c48c7904c8b5c756370affbe5a5fb8bb08eadd0480f0a3f1aed925c0c1abaa) to point to the correct jsoo version.

Checklist:

- [x] Dependency versions are unchanged
  - Notify Velocity team if dependencies must change in CI
- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them

* Closes #0000

### CODE REVIEW

**Comment by dannywillems** [comment]
!ci-build-me

**Comment by dannywillems** [comment]
!ci-build-me

File: scripts/testone.sh
Comment (rbonichon) [inline scripts/testone.sh:1]: This change might belong to another PR (it's not related to jsoo/alcotest/dune)

File: src/lib/mina_metrics/prometheus_metrics/mina_metrics.ml [hunk 1]
```diff
@@ -81,7 +81,7 @@ module TextFormat_0_0_4 = struct
       | None ->
           (label_names, label_values)
       | Some (label_name, label_value) ->
-          let label_value_str = Fmt.strf "%a" output_value label_value in
+          let label_value_str = Fmt.str "%a" output_value label_value in
```
Comment (rbonichon) [inline src/lib/mina_metrics/prometheus_metrics/mina_metrics.ml:84]: Since you're changing this line, you might as well use
```suggestion
          let label_value_str = Format.sprintf "%a" output_value label_value in
```

File: src/lib/pickles/tests/test_ro.ml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+(* initially written to see the consistency between updates of digestif *)
+let test_bits_random_oracle_consistency_check () =
+  let s = "BitsRandomOracle" in
+  let exp_output_str =
+    "01000001101110001111111100011000011000100010100011010010110110010011101101101000001110001110100101010100001000001110101110111010"
+  in
+  let exp_output =
+    List.map (fun i -> i = '1') (List.of_seq (String.to_seq exp_output_str))
```
Comment (rbonichon) [inline src/lib/pickles/tests/test_ro.ml:8]:
```suggestion
    List.map (fun i -> i = '1') (String.to_list exp_output_str)
```
Or, for OCaml golfer (see [here](https://www.vimgolf.com/) for the reference to golf) who wants to spare iterations
```suggestion
    let l = ref [] in 
    for i = String.length exp_output_str - 1 downto 0 do
        l := exp_output_str.(i) = '1' :: !l
    done;
    !l
```
:-)

File: src/lib/pickles/tests/test_ro.ml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  in
+  let output = Pickles.Ro.bits_random_oracle ~length:128 s in
```
Comment (rbonichon) [inline src/lib/pickles/tests/test_ro.ml:10]: We don't really need to export `Ro` as part of the interface for Pickles to test it. We could use `Pickles__Ro` as well. 

I'm not sure this is per se a good practice to do that, nor is it probably to expose a module of a library that is, as far as it looks, purely local.

Was it necessary to update `alcotest` to run this test?

**Comment by dannywillems** [comment]
!ci-build-me

File: src/lib/pickles/tests/test_ro.ml (hunk 2 above)
Comment (dannywillems) [reply src/lib/pickles/tests/test_ro.ml:8]: Oh, it seems Core_kernel implements the to_list function in the String module. That's nice! Gonna start liking Core...

Comment (dannywillems) [reply src/lib/pickles/tests/test_ro.ml:8]: For a test, I prefer to stay functional :)

File: src/lib/pickles/tests/test_ro.ml (hunk 3 above)
Comment (dannywillems) [reply src/lib/pickles/tests/test_ro.ml:10]: No, it was not necessary to update alcotest for this test.
It is part of testing the update to digestif, in addition to adding a small test.

I agree with your points regarding exposong `Ro`. I'm not in favor of using `Pickles__Ro`, and I am more used to expose the module to test it. Fine doing it if you want.

There is a way to bypass it by exposing a module `Pickles_internal` to unit test functions if we don't want to fully exposed Ro in the API.

Another way is to use inline tests. However, I remember we were thinking about moving to alcotest.

**Comment by dannywillems** [comment]
!ci-bulid-me

**Comment by dannywillems** [comment]
!ci-build-me

**Comment by dannywillems** [comment]
!ci-build-me

File: src/lib/pickles/tests/test_ro.ml (hunk 3 above)
Comment (rbonichon) [reply src/lib/pickles/tests/test_ro.ml:10]: Well. I prefer moving to `alcotests`, mostly for engineering reasons. Shall we make an issue of exposing `Pickles_internal` then?

File: src/lib/pickles/tests/test_ro.ml (hunk 2 above)
Comment (rbonichon) [reply src/lib/pickles/tests/test_ro.ml]: Then rewrite the for loop as 
```
let rec loop acc i = if i < 0 then acc else loop (exp_output_str.(i) = '1' :: acc) (i - 1) in 
loop [] (String.length exp_output_str - 1)
```
;-)

In any case, it's high-level nitpicking since it's just a single test.

File: scripts/testone.sh
Comment (rbonichon) [reply scripts/testone.sh:1]: That's sad

**Comment by dannywillems** [comment]
!ci-build-me

**Comment by dannywillems** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
Closing as stale. This PR is referenced in https://github.com/MinaProtocol/mina/issues/14829, so that it can be used as a starting point if / when work resumes on this.
